sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/bigquery/BigQueryExecutor.java

```diff
@@ -0,0 +1,332 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryExecutor.class);
+
+  public BigQueryExecutor(SQLConfig config, Map<String, SQLTransform> pluginMap) {
+    this.config = config;
+    this.sourceStages = new HashSet<>();
+    this.sinkStages = new HashSet<>();
+    this.pluginMap = pluginMap;
+    this.fromNodeMap = new HashMap<>();
+    this.toNodeMap = new HashMap<>();
+    this.relationalNodeMap = new HashMap<>();
+    this.temporaryTableMap = new HashMap<>();
+    this.queries = new ArrayList<>();
+  }
+
+  @Override
+  public void configure(CustomActionConfigurer configurer) {
+    // Initialize the query generation
+    init();
+
+    // Store the config JSON for access during runtime
+    Gson gson = new GsonBuilder().create();
+    Map<String, String> properties = new HashMap<>();
+    properties.put(CONFIG_NAME, gson.toJson(config));
+    configurer.setProperties(properties);
+  }
+
+  @Override
+  public void initialize(CustomActionContext context) {
+    // Initialize all of the instance variables
+    sourceStages = new HashSet<>();
+    sinkStages = new HashSet<>();
+    pluginMap = new HashMap<>();
+    fromNodeMap = new HashMap<>();
+    toNodeMap = new HashMap<>();
+    relationalNodeMap = new HashMap<>();
+    temporaryTableMap = new HashMap<>();
+    queries = new ArrayList<>();
+
+    // Set the config
+    Gson gson = new GsonBuilder().create();
+    config = gson.fromJson(context.getSpecification().getProperty(CONFIG_NAME), SQLConfig.class);
+
+    for (ETLStage stage: config.getStages()) {
+      String node = stage.getName();
+
+      // Instantiate the plugin
+      SQLTransform plugin;
+      try {
+        plugin = context.newPluginInstance(node);
+      } catch (InstantiationException e) {
+        // This probably should have been caught at configure time
+        throw new IllegalStateException("Failed to instantiate plugin for stage " + node);
+      }
+      pluginMap.put(node, plugin);
+    }
+
+    // Initialize the query generation
+    init();
+  }
+
+  @Override
+  public void run() throws Exception {
+    StringBuilder queryBuilder = new StringBuilder();
+    // Construct queries
+    for (String query: queries) {
+      queryBuilder.append(query);
+      queryBuilder.append(";\n");
+    }
+
+    String queries = queryBuilder.toString();
+    LOG.info("Executing queries: " + queries);
+    GoogleCredentials credentials;
+    File credentialsPath = new File(serviceAccountPath);
+    FileInputStream serviceAccountStream = new FileInputStream(credentialsPath);
+    credentials = ServiceAccountCredentials.fromStream(serviceAccountStream);
+
+    // Instantiate a client.
+    BigQuery bigquery =
+      BigQueryOptions.newBuilder().setCredentials(credentials).build().getService();
+    QueryJobConfiguration queryConfig = QueryJobConfiguration.newBuilder(queries).build();
+    bigquery.query(queryConfig);
+  }
+
+  @Override
+  public void destroy() {}
+
+  private void init() {
```

Comment:
This method is too long. Can you break down this to several helper functions? Like load the driver can be one function, compute the connection and stages can be one function, etc. With some javadoc, it is much clearer to read the code.

Reply:
Fixed, split the function up and added some javadocs